File: 6o Semester/Artificial-Intelligence/hw1b/FreeCellSolitaire/Algorithms.py
import time
timeLimit=300


def breadthFirstSearch(root):
    print('Breadth First search...')
    start=time.time()
    queue=[root]
    visitedList=list()
    visitedSet=set()
    movesToFinish=list()
    loops=0
    executionTime=time.time()-start

    while len(queue) > 0 and executionTime < timeLimit:

        currentNode=queue.pop(0)
        movesToFinish.append(currentNode.move)

        if currentNode==root:
            visitedList.append(currentNode)
            visitedSet.add(currentNode.valueToStr())

        if currentNode.checkIfWin():
            executionTime=time.time()-start
            return movesToFinish,executionTime,True

        allExpandMoves=currentNode.expandNode()

        for index,child in enumerate(currentNode.children):
            if not nodeAlreadyExists(visitedSet, child):
                queue.append(child)
                visitedList.append(child)
                visitedSet.add(child.valueToStr())
                child.move=allExpandMoves[index]

        loops+=1
        executionTime=time.time()-start

    print('The BFS algorithm failed to solve the problem within 5 minutes.')
    return movesToFinish,executionTime,False


def depthFirstSearch(root):
    print('Depth First search...')
    start=time.time()
    stack=[root]
    childrenToBeAdded=list()
    visitedList=list()
    visitedSet=set()
    movesToFinish=list()
    loops=0
    executionTime=time.time()-start

    while len(stack) > 0 and executionTime < timeLimit:
        
        currentNode=stack.pop(0)
        childrenToBeAdded.clear()
        movesToFinish.append(currentNode.move)

        if currentNode==root:
            visitedList.append(currentNode)
            visitedSet.add(currentNode.valueToStr())
        
        if currentNode.checkIfWin():
            executionTime=time.time()-start
            return movesToFinish,executionTime,True

        allExpandMoves=currentNode.expandNode()

        for index, child in enumerate(currentNode.children):
            if not nodeAlreadyExists(visitedSet, child):
                childrenToBeAdded.append(child)
                visitedList.append(child)
                visitedSet.add(child.valueToStr())
                child.move=allExpandMoves[index]

        childrenToBeAdded.reverse()
        for child in childrenToBeAdded:
            stack.insert(0,child)

        loops+=1
        executionTime=time.time()-start

    print('The DFS algorithm failed to solve the problem within 5 minutes.')
    return movesToFinish,executionTime,False


def bestFirstSearch(root):
    print('Best First search...')
    start=time.time()
    notVisitedList=[root]
    visitedList=list()
    visitedSet=set()
    movesToFinish=list()
    loops=0
    executionTime=time.time()-start

    while len(notVisitedList) > 0 and executionTime < timeLimit:

        currentNode=notVisitedList.pop(0)
        movesToFinish.append(currentNode.move)

        if currentNode==root:
            visitedList.append(currentNode)
            visitedSet.add(currentNode.valueToStr())
        
        if currentNode.checkIfWin():
            executionTime=time.time()-start
            return movesToFinish,executionTime,True

        allExpandMoves=currentNode.expandNode()

        for index,child in enumerate(currentNode.children):
            if not nodeAlreadyExists(visitedSet, child):
                notVisitedList.append(child)
                visitedList.append(child)
                visitedSet.add(child.valueToStr())
                child.move=allExpandMoves[index]

        notVisitedList.sort(key=lambda Node: Node.bestFirstScore)

        loops+=1
        executionTime =time.time()-start

    print('The algorithm best first search failed to solve the problem within 5 minutes.')
    return movesToFinish,executionTime,False


def ASearch(root):
    print('A* search...')
    start=time.time()
    notVisitedList=[root]
    visitedList=list()
    visitedSet=set()
    movesToFinish=list()
    loops=0
    executionTime=time.time()-start

    while len(notVisitedList) > 0 and executionTime < timeLimit:

        currentNode=notVisitedList.pop(0)
        movesToFinish.append(currentNode.move)

        if currentNode==root:
            visitedList.append(currentNode)
            visitedSet.add(currentNode.valueToStr())

        if currentNode.checkIfWin():
            executionTime=time.time()-start
            return movesToFinish,executionTime,True

        allExpandMoves=currentNode.expandNode()

        for index,child in enumerate(currentNode.children):
            if not nodeAlreadyExists(visitedSet, child):
                notVisitedList.append(child)
                visitedList.append(child)
                visitedSet.add(child.valueToStr())
                child.move=allExpandMoves[index]

        notVisitedList.sort(key=lambda Node: Node.AScore)
        
        loops+=1
        executionTime=time.time()-start

    print('The Α* algorithm failed to solve the problem within 5 minutes.')
    return movesToFinish,executionTime,False




def nodeAlreadyExists(visitedSet,child):
    return child.valueToStr() in visitedSet

File: 6o Semester/Artificial-Intelligence/hw1c/Knuth/Algorithms.java
import java.util.ArrayList;
import java.util.Stack;

public class Algorithms {
	static long timeLimit1=60000;
	static long timeLimit2=30000;
	
	
	public static ArrayList<String> breadthFirstSearch(Tree root) {
		long start=System.currentTimeMillis();
		
		ArrayList<String> moves=new ArrayList<String>();
		ArrayList<Tree> queue=new ArrayList<Tree>();
		queue.add(root);
		long executionTime=System.currentTimeMillis()-start;
		
		while(queue.size()>0 && executionTime<=timeLimit1) {
			
			Tree currentNode=queue.remove(0);
			moves.add(currentNode.move);
			
			if (currentNode.win()){
				executionTime=System.currentTimeMillis()-start;
				String time=String.valueOf(executionTime);
				moves.add(time);
				return moves;
			}
			
			currentNode.expandNode();
			for(Tree child: currentNode.children) {
				queue.add(child);
			}
			executionTime=System.currentTimeMillis()-start;
			
			
		}
		System.out.println("The BFS algorithm failed to solve the problem within 60 seconds.");
		System.out.println("Number of movements:"+String.valueOf(moves.size()-1));
		return new ArrayList<String>();
		
	}
	
	
	
	
	public static ArrayList<String> iterativeDeepening(Tree root){
		long start=System.currentTimeMillis();
		ArrayList<String> moves=new ArrayList<String>();
		int depth_limit=1;
		long executionTime=System.currentTimeMillis()-start;
		boolean win=true;
		
		while(win && executionTime<=timeLimit2)
		{
			
			moves=depthFirstSearch(root,depth_limit,executionTime,start);
			if(moves.size()>0)
			{
				
				win=false;
				return moves;
			}
			
			depth_limit++;
			executionTime=System.currentTimeMillis()-start;
			
		}
		
		System.out.println("The ID algorithm failed to solve the problem within 30 seconds.");
		return new ArrayList<String>();
	}
	
	
	
	public static ArrayList<String> depthFirstSearch(Tree root,int depth_limit,long executionTime,long start){
		ArrayList<String> moves=new ArrayList<String>();
		Stack<Tree> stack=new Stack<Tree>();
		stack.push(root);
		
		
		while(stack.size()>0 && executionTime<=timeLimit2)
		{
			Tree currentNode=stack.pop();
			moves.add(currentNode.move);
			
			
			if (currentNode.win()){
				executionTime=System.currentTimeMillis()-start;
				String time=String.valueOf(executionTime);
				moves.add(time);
				return moves;
			}
			
			if(currentNode.depth<=depth_limit && currentNode.depth!=depth_limit) 
			{
				currentNode.expandNode();
				
				for(Tree child: currentNode.children) {
					stack.push(child);
					}	
			}
			
			executionTime=System.currentTimeMillis()-start;
			
		}
		
		return new ArrayList<String>();
	}
	
	
		
	


}
